README.md
# pysession
**pysession** is a Python module for establishing sessions to network devices via ssh, telnet or console, signing in with provided login credentials, detecting device prompt, sending command and retrieving output. 

pysession is based on pexpect, so in other words, it is an extension of pexpect. It provides a simplified interface to access and controle network devices by hidding the lower-level interaction, like
* Login device (detecting prompts for user id, password and enable password, and handling accordingly)
* Automatic device prompt detection during login process or after special commands like (config term, rconsole...)
* Page break (either entering page-break-free mode, or feeding space if seeing page-break)
* Self-defined prompts and handling. (please see the below section for details

So as a user, you don't need to spend your efforts/time on the script hundreds of lines code to deal with router interaction, instead just focus on what commands to send and what output returns. 

## How to use
You can use pysessin either as a standalone tool or a python library for your own python program. 

* Standalone tool, actually i use this way to test the library after making any changes
```
./pysession.py -s 'telnet 1.1.1.1 2001; ssh admin@1.1.1.2' -c 'show ver; show ip route' -p pswd -e enablepswd'
```
Then this python script will login the sessions of -s paramenter and run the commands specified in -c argument. 

* Python library
```
import pysession

rtr = pysession(session='telnet 10.1.1.1', user="admin", password="password")

output = rtr.send('''
show ver
show run
show ip route
''')
```
## How to install
Just copy the pysession.py and pysession.conf to local directory. (there should be a better/professional way, but I will deal with that later)


## Self-defined prompts and handling
If your devices have unique/special/weird prompts, you don't need to modify the script to handle them, instead just change the "pysession.conf" file. 

Example #1, the device login prompt to ask for username is not "login:" or "username:", instead it is "Tell me who are you?". 

You can edit pysession.conf file to add one line under 
```
[Prompt.Stage.Login]
Tell me who are you? = $user
```

Example #2, the terminal server asks if you want to clear the existing session
```
!!! tty is being used !!!
1 - Initiate a regular session
2 - Quit
Enter your option :
```
You just need to add one line like under [Prompt.Stage.Login]
```
[Prompt.Stage.Login]
1 - Initiate a regular session.*Enter your option : = 1$   
```
_the ending $ means only sending "1" without return._



CHANGELOG.md
#version 0.1
* Initial version
* Device login
* prompt parse
* 3 access support - telnet, ssh, console. and test

#version 0.2
* page-break
* prompt parse for config mode

#version 1.0 
* pysession.conf - user-defined prompts/handling instead of hard coded in script 
* jump server - support command based jump server

#version 2.0 - To Do
* router script


examples/PyBrcdDataCollector.py
#!/usr/bin/env python

import re
import sys
import time
from pysession import *

def parse_sh_ip_route_prefix(output):
    """
    telnet@CER1-solomon-1464141#show ip route 3.1.10.3
    Type Codes - B:BGP D:Connected I:ISIS O:OSPF R:RIP S:Static; Cost - Dist/Metric
    BGP  Codes - i:iBGP e:eBGP
    ISIS Codes - L1:Level-1 L2:Level-2
    OSPF Codes - i:Inter Area 1:External Type 1 2:External Type 2 s:Sham Link
    STATIC Codes - d:DHCPv6 
            Destination        Gateway         Port          Cost          Type Uptime src-vrf 
    1       3.1.10.0/24        12.1.10.2       ve 10         1/1           S    14h50m -

    """
    reach_content = False

    for line in output.split('\n'):
        if re.search("^\s+Destination\s+Gateway\s+Port.*", line):
            reach_content = True
            continue

        if not reach_content:
            continue

        # now the line is like:
        # 1       3.1.10.0/24        12.1.10.2       ve 10         1/1           S    14h50m -
        #if re.search("^\d+\s+([0-9\.\/]+)\s+([0-9\.\/]+)", line): 
        return map(str.strip, re.split('\s\s+', line))
    
    return [''] * 8

def ipv4_addr_full_format(ip_addr):
    """
    3.1.1.3 -> 003.001.001.003
    """
    ip_addr = ip_addr.strip()

    re_ipv4_found = re.search('^(\d+)\.(\d+)\.(\d+)\.(\d+)$', ip_addr)

    if re_ipv4_found:
        l = map(int, re_ipv4_found.groups())
        return '%03d.%03d.%03d.%03d' % (l[0], l[1], l[2], l[3])

    return ''

def netmask_to_dot_notation(netmask):
    """
    netmask /n to 255.255.255.0
    """
    netmask = int(netmask)

    dot_mask_list = []
    for i in range(4):
        if netmask >= 8:
            dot_mask_list.append('255')
        elif netmask <= 0:
            dot_mask_list.append('0')
        else:
            dot_mask_list.append(str(2**netmask))

        netmask -= 8
    
    return '.'.join(dot_mask_list)
    
def parse_dm_metro_rc_for_NH_index(output, network, mask_dot):
    """
LP-1#dm metro 0 rc l3 | include 3.1.30.0|NH
|     CAM      |   VPRAM    |VRF|  IP  Addr     |  IP Mask       | NH   | Arp  | A|T
166263;  14984; 0x02000985 ;  0;        3.1.30.0;  255.255.255.0;  2437;   211| 0| 0
    """
    for line in output.split('\n'):
        re_nh_found = re.search('%s\;\s+%s\;\s+(\d+)\;' % (issue_prefix, net_mask_dot), line)

        if re_nh_found:
            return re_nh_found.group(1)

    return ''
    
if __name__ == '__main__': 

    # get the issue prefix
    #issue_dest = '3.1.30.3'
    print '''
###############################################################################
#                                                                             #
#                         Brocade Data Collection Tool                        #
#                                   ver 0.1                                   #
#                                                                             #
#                          Author: solomon@brocade.com                        #
#                                                                             #
###############################################################################
    
    '''

    issue_dest = raw_input('please provide the dest ip of issue prefix: ')

    # get the session info interactively
    cer = pysession()
    #cer = pysession(session='telnet 10.18.24.78')
    cer.pprint=True

    # Part 0. basic info
    o = cer.send("show ip route %s" % issue_dest)

    dummy, issue_prefix, next_hop, oif, cost, prefix_type, prefix_uptime, vrf\
        = parse_sh_ip_route_prefix(o)[:8]

    # get netmask in dot notation
    issue_prefix, net_mask = issue_prefix.split('/')
    net_mask_dot = netmask_to_dot_notation(net_mask)

    print '\n' + pys_lib.pline2('Basic Prefix Info')

    pys_lib.pys_pprint(\
        ['Issue Dest', 'Issue Prefix', 'Net Mask', 'Net Mask Dot', \
         'Next Hop', 'OutIntf', 'Cost', 'Prefix Type', 'Prefix Uptime', \
         'Prefix VRF'],
        [issue_dest, issue_prefix, net_mask, net_mask_dot, next_hop, oif, 
         cost, prefix_type, prefix_uptime, vrf])

    # Part 1, MP commands
    print '\n' + pys_lib.pline2('Part 1, MP commands')

    cer.send("skip")
    cer.send("show clock")
    cer.set_debug_dest_to_me()
    cer.send("show tech-support", timeout=60)
    #cer.send("ping %s" % issue_dest, timeout=30)
    cer.send("dm pstat")
    cer.send("show arp %s debug" % next_hop)
    cer.send("show ip arp-mac-entries")
    pys_lib.psleep(120, ' 2nd dm pstat')
    cer.send("dm pstat")

    # Part 2, LP commands
    print '\n' + pys_lib.pline2('Part 2, LP commands')

    cer.send("rconsole 1")
    cer.send("enable")
    cer.send("skip")
    cer.send("show ip arp-mac")
    cer.send("show arp %s debug" % next_hop)
    cer.send("show ip cache")
    cer.send("show ip cache %s" % issue_prefix)
    cer.send("show ip route", timeout=180)
    cer.send("show ip route %s" % issue_prefix)
    cer.send("show ip next-hop %s" % issue_prefix)
    cer.send("show ip next-hop %s debug" % issue_prefix)
    cer.send("dm metro 0 lpm | include %s" % ipv4_addr_full_format(issue_prefix))
    cer.send("show ip lpm debug %s" % issue_prefix)

    output = cer.send("dm metro 0 rc l3 | include %s" % issue_prefix, timeout=300)
    nh_index = parse_dm_metro_rc_for_NH_index(output, issue_prefix, net_mask_dot)

    cer.send("dm metro 0 table nexthop %s detail" % nh_index)

    # Part 3, DEBUG commands
    print '\n' + pys_lib.pline2('Part 3, DEBUG commands')

    cer.send("debug ip next-hop")
    cer.send("debug ip arp %s" % next_hop)
    cer.send("debug ip arp-mac")
    cer.send("debug ip rtm %s" % issue_prefix)
    cer.send("debug ip cache")
    cer.send("debug ip lpm")
    cer.send("debug ip static-cam")

    cer.send("debug ip rtm nexthop")
    
    pys_lib.psleep(300, 'waiting for debug output')
    cer.send("\n")
    #print '\n\n' + '-' * 80 + '\n' + o + '\n' + '=' * 80

    pys_lib.psleep(300, 'waiting for debug output')
    cer.send("\n")
    #print '\n\n' + '-' * 80 + '\n' + o + '\n' + '=' * 80

    pys_lib.psleep(300, 'waiting for debug output')
    o = cer.send("\n")
    #print '\n\n' + '-' * 80 + '\n' + o + '\n' + '=' * 80

    cer.send("no debug all")
    cer.send("exit")
    cer.send("no debug all")
    cer.send("exit")
    cer.send("no debug all")
    
    print '\n' + pys_lib.pline2('END OF DATA COLLECTION')


tools/BrcdDataCollector.py
#!/usr/bin/env python

import re
import os
import sys
import time
import getopt
from pysession import *

def read_cmd_file(cmdfile):
    list_cmd = []

    with open(cmdfile) as f:
        for line in f:
            line1 = line.strip()

            if line1 == '' or line1[0]=='#':
                continue

            list_cmd.append(line.rstrip())

    return '\n'.join(list_cmd)

def usage():
    print '''
Usage: BrcdDataCollector.py -C <cmdfile> -s <session> -i <userid> [-p <password>] [-e <enable_password]

arguments:
    -C, --cmdfile          command file containing list of commands send to router
    -s, --session          session info like "telnet 10.1.1.1" or "ssh -l admin gw1.company.com"
    -u, --userid           user id
    -p, --password         login password of user id (you can leave it blank and provide it later in non-echo way)
    -e, --enable_password  enable password (you can leave it blank and provide it later in non-echo way)
    -o, --output           output file name
    '''

def print_header():
    print r'''
/##############################################################################\
#                          Brocade Data Collection Tool                        #
#                                    ver 0.1                                   #
#                           Author: solomon@brocade.com                        #
\##############################################################################/
    '''

if __name__ == '__main__': 
    try:
        opts, args = getopt.getopt(sys.argv[1:], "C:S:s:u:p:e:o:", 
            ["cmdfile=", "sessfile=", "session=", "userid=", "password=",\
             "enable_password=", "output="])
    except getopt.GetoptError:
        usage()
        sys.exit(2)

    # initialize the variables
    all_commands = ''
    session, userid, password, enable_password, output_file = '', '', '', '', ''

    # parse the sys.argv
    for opt, arg in opts:
        if opt in ('-h', '--help'):
            usage()
        elif opt in ('-C', '--cmdfile'):
            all_commands = read_cmd_file(arg)
        elif opt in ('-S', '--sessfile'):
            sessfile = arg
            # not completed yet
        elif opt in ('-s', '--session'):
            session = arg
        elif opt in ('-u', '--userid'):
            userid = arg
        elif opt in ('-p', '--password'):
            password = arg
        elif opt in ('-o', '--output'):
            output_file = arg
        elif opt in ('-e', '--enable_password'):
            enable_password = arg

    if all_commands == '':
        print "Error: no commands specified!!!!!"
        usage()
        sys.exit(2)

    print_header()

    print '\n' + PYSLib.pline1('SESSION INFO')
    print "Session: %s" % session
    print "User ID: %s" % userid
    print PYSLib.pline1('Press Enter to continue ......')
    dummy = raw_input("")

    print '\n' + PYSLib.pline1('LIST OF COMMANDS')
    print all_commands
    print PYSLib.pline1('Press Enter to start......')
    dummy = raw_input("")

    router = pysession(session=session, user=userid, password=password, 
        enable_password=enable_password, log_file_prefix='brcd__', 
        output_file=output_file)

    router.pprint=True 
    
    router.send(all_commands)

    total_sec = int(time.time() - router.start_time)
    elapse_time = '%d min %d sec' % (total_sec/60, total_sec%60)

    sleep_time = '%d min %d sec' % (router.sleep_time/60, router.sleep_time%60)

    file_size_in_KB = '%d KB' % int(os.stat(router.log_file_name).st_size/1000)

    print '\n' + PYSLib.pline2('DATA COLLECTION SUMMARY')
    print '       script running time :', elapse_time
    print '             sleeping time :', sleep_time
    print '           number of lines :', router.counter_line
    print '        number of commands :', router.counter_cmd
    print 'number of invalid commands :', router.counter_invalid_cmd
    print '   size of log output file :', file_size_in_KB
    print PYSLib.pline2('END OF DATA COLLECTION')
    
